### ISSUE TITLE
Segmented Fusion hitting root domain assertions

### ISSUE THREAD

**Issue by shmsong** [csarofeen/pytorch#709]
## 🐛 Bug

This is an issue tracking a failure identified by segmented bachnorm kernel. 

## To Reproduce

```cpp
TEST(NVFuserTest, FusionBNRepro_CUDA) {
  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
  Fusion& fusion = *fusion_ptr.get();
  FusionGuard fg(&fusion);

  int batch = 4;
  int c = 4;
  int h = 4;
  int w = 4;
  int numDims = 4;

  auto input = makeSymbolicTensor(numDims);
  fusion.addInput(input);
  auto weight = makeSymbolicTensor(1);
  fusion.addInput(weight);
  auto bias = makeSymbolicTensor(1);
  fusion.addInput(bias);
  auto running_mean = makeSymbolicTensor(1);
  fusion.addInput(running_mean);
  auto running_var = makeSymbolicTensor(1);
  fusion.addInput(running_var);

  // TODO: error set 1, runtime momentum;
  Val* momentum_ptr = new Double(0.1);
  Val* rev_momentum_ptr = new Double(1.0 - 0.1);
  Val* eps_ptr = new Double(1e-5);

  std::vector<int> reduction_axes;
  std::vector<bool> broadcast_mask(numDims, false);
  Val* num_features = new Double(1);
  for (size_t axis = 0; axis < numDims; ++axis) {
    if (axis != 1) {
      reduction_axes.push_back(axis);
      broadcast_mask[axis] = true;
      num_features = mul(
          num_features, input->domain()->domain()[axis]->extent());
    }
  }

  // Algorithm
  auto x_sum = sum(input, reduction_axes);
  auto x_sum_bcast = broadcast(x_sum, broadcast_mask);
  auto x_mean = div(x_sum_bcast, num_features);
  
  // updating running mean
  auto current_mean_hat = mul(x_mean, momentum_ptr);
  auto mean_hat = mul(running_mean, rev_momentum_ptr);
  auto new_mean_hat = add(mean_hat, current_mean_hat);
  fusion.addOutput(new_mean_hat);
  //fusion->aliasOutputToInput(new_mean_hat, running_mean);
  
  auto x_mean_sub = sub(input, x_mean);
  auto x_mean_sub_pow = mul(x_mean_sub, x_mean_sub);
  auto var_sum = sum(x_mean_sub_pow, reduction_axes);
  auto var_sum_bcast = broadcast(var_sum, broadcast_mask);
  auto var = div(var_sum_bcast, num_features);
  
  
  // updating running var
  auto num_feature_decrement = sub(num_features, new Int(1));
  auto unbiased_var = div(var_sum, num_feature_decrement);
  auto current_var_hat = mul(unbiased_var, momentum_ptr);
  auto var_hat = mul(running_var, rev_momentum_ptr);
  auto new_var_hat = add(var_hat, current_var_hat);
  fusion.addOutput(new_var_hat);
  //fusion->aliasOutputToInput(new_var_hat, running_var);
  
  auto var_eps = add(var, eps_ptr);
  auto invstd = unaryOp(UnaryOpType::Rsqrt, var_eps);
  auto output = mul(x_mean_sub, invstd);
  
  // Optional: norm * weight
  if (weight) {
    auto weight_bcast = broadcast(weight, broadcast_mask);
    output = mul(output, weight_bcast);
  }
  if (bias) {
    auto bias_bcast = broadcast(bias, broadcast_mask);
    output = add(output, bias_bcast);
  }
  fusion.addOutput(output);
  auto save_mean = sum(x_mean, reduction_axes);
  fusion.addOutput(save_mean);
  auto save_invstd = sum(invstd, reduction_axes);
  fusion.addOutput(save_invstd);

  fusion.printMath();

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::Tensor input1 = at::randn({batch, c, h, w}, options);
  at::Tensor input2 = at::randn({c}, options);
  at::Tensor input3 = at::randn_like(input2);
  at::Tensor input4 = at::randn_like(input2);
  at::Tensor input5 = at::randn_like(input2);

  FusionExecutorCache fec(std::move(fusion_ptr));
  fec.fusionSegments()->print();
  std::vector<IValue> inputs = {input1, input2, input3, input4, input5};
  auto outputs = fec.runFusionWithInputs(inputs);
}
```

**Comment by shmsong** [comment]
Error message: 
```
C++ exception with description "in_root.size() == out_root.size()INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/root_domain_map.cpp":643, please report a bug to PyTorch. 
Exception raised from mapPointwiseOrReductionOp at ../torch/csrc/jit/codegen/cuda/root_domain_map.cpp:643 (most recent call first):
frame #0: <unknown function> + 0xda326 (0x7f3df329e326 in /opt/home/pytorch_debug/build/lib/libc10.so)
frame #1: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > ()>::operator()() const + 0x54 (0x7f3dc0b7a1a4 in /opt/home/pytorch_debug/build/lib/libtorch_cpu.so)
frame #2: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x57 (0x7f3df329d747 in /opt/home/pytorch_debug/build/lib/libc10.so)
frame #3: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x9f (0x7f3df329bdbf in /opt/home/pytorch_debug/build/lib/libc10.so)
frame #4: bin/test_jit() [0x547c3b]
frame #5: torch::jit::fuser::cuda::ComputeAtRootDomainMapBuilder::mapPointwiseOrReductionOp(torch::jit::fuser::cuda::Expr*) + 0x249 (0x7f3df2927ca9 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #6: <unknown function> + 0x2fa145c0 (0x7f3df292c5c0 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #7: void torch::jit::fuser::cuda::Expr::dispatch<torch::jit::fuser::cuda::OptOutDispatch*>(torch::jit::fuser::cuda::OptOutDispatch*, torch::jit::fuser::cuda::Expr*) + 0xda (0x7f3df27394ca in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #8: torch::jit::fuser::cuda::OptOutDispatch::handle(torch::jit::fuser::cuda::Expr*) + 0x1d (0x7f3df2737bed in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #9: torch::jit::fuser::cuda::BackwardVisitor::handle(torch::jit::fuser::cuda::Expr*) + 0x20 (0x7f3df282b5a0 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #10: torch::jit::fuser::cuda::ComputeAtRootDomainMapBuilder::handle(torch::jit::fuser::cuda::Expr*) + 0x83 (0x7f3df2927a33 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #11: void torch::jit::fuser::cuda::Statement::dispatch<torch::jit::fuser::cuda::OptOutDispatch*>(torch::jit::fuser::cuda::OptOutDispatch*, torch::jit::fuser::cuda::Statement*) + 0x93 (0x7f3df27382b3 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #12: torch::jit::fuser::cuda::OptOutDispatch::handle(torch::jit::fuser::cuda::Statement*) + 0x1d (0x7f3df2737bbd in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #13: torch::jit::fuser::cuda::BackwardVisitor::handle(torch::jit::fuser::cuda::Statement*) + 0x20 (0x7f3df282b570 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #14: torch::jit::fuser::cuda::BackwardVisitor::traverseFrom(torch::jit::fuser::cuda::Fusion*, std::vector<torch::jit::fuser::cuda::Val*, std::allocator<torch::jit::fuser::cuda::Val*> > const&, bool) + 0x7c6 (0x7f3df282bda6 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #15: torch::jit::fuser::cuda::ComputeAtRootDomainMapBuilder::ComputeAtRootDomainMapBuilder(torch::jit::fuser::cuda::ComputeAtRootDomainMap&, bool) + 0x475 (0x7f3df2926d55 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #16: torch::jit::fuser::cuda::ComputeAtRootDomainMap::build(bool) + 0x7a (0x7f3df2924eea in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #17: <unknown function> + 0x2f802055 (0x7f3df271a055 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #18: <unknown function> + 0x2f7ffa02 (0x7f3df2717a02 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #19: torch::jit::fuser::cuda::TensorView::computeAt(torch::jit::fuser::cuda::TensorView*, int, torch::jit::fuser::cuda::ComputeAtMode) + 0x119 (0x7f3df2977369 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)
frame #20: torch::jit::fuser::cuda::scheduleNormalization(torch::jit::fuser::cuda::Fusion*, torch::jit::fuser::cuda::ReductionParams const&, std::vector<torch::jit::fuser::cuda::TensorView*, std::allocator<torch::jit::fuser::cuda::TensorView*> > const&, std::vector<torch::jit::fuser::cuda::TensorView*, std::allocator<torch::jit::fuser::cuda::TensorView*> >&) + 0x2823 (0x7f3df29405e3 in /opt/home/pytorch_debug/build/lib/libtorch_cuda.so)

```

**Comment by shmsong** [comment]
This is an issue with using `cache_fork` in `scheduleNormalization`. Kernel math of the segmented fusion is:
```
%kernel_math {
T5[ iS9{i3}, rS100{( ( i1 * i5 ) * i7 )} ] = reduction( T0[ iS1{i3}, iS94{( ( i1 * i5 ) * i7 )} ], op = add, initial value = double(0) )
T6[ iS13{i3}, bS116{( ( 1 * 1 ) * 1 )} ] = broadcast( T5[ iS9{i3}, rS100{( ( i1 * i5 ) * i7 )} ] )
d20 = double(1) * i1;
d21 = d20 * i5;
d22 = d21 * i7;
T7[ iS16{i3}, bS122{( ( 1 * 1 ) * 1 )} ]
   = T6[ iS13{i3}, bS116{( ( 1 * 1 ) * 1 )} ]
   / d22;
T8[ iS20{i3}, bS114{( ( 1 * 1 ) * 1 )} ]
   = T7[ iS16{i3}, bS122{( ( 1 * 1 ) * 1 )} ]
   * double(0.10000000000000001);
T11[ iS30{i3}, iS130{( ( 1 * 1 ) * i13 )} ]
   = T10[ bS26{1}, iS96{( ( 1 * 1 ) * i13 )} ]
   + T8[ iS20{i3}, bS114{( ( 1 * 1 ) * 1 )} ];
T28[ iS86{i3}, rS124{( ( 1 * 1 ) * 1 )} ] = reduction( T7[ iS16{i3}, bS122{( ( 1 * 1 ) * 1 )} ], op = add, initial value = double(0) )
T12[ iS34{i3}, iS120{( ( i1 * i5 ) * i7 )} ]
   = T0[ iS1{i3}, iS94{( ( i1 * i5 ) * i7 )} ]
   - T7[ iS16{i3}, bS122{( ( 1 * 1 ) * 1 )} ];
T13[ iS38{i3}, iS126{( ( i1 * i5 ) * i7 )} ]
   = T12[ iS34{i3}, iS120{( ( i1 * i5 ) * i7 )} ]
   * T12[ iS34{i3}, iS120{( ( i1 * i5 ) * i7 )} ];
T14[ iS42{i3}, rS98{( ( i1 * i5 ) * i7 )} ] = reduction( T13[ iS38{i3}, iS126{( ( i1 * i5 ) * i7 )} ], op = add, initial value = double(0) )
T15[ iS46{i3}, bS128{( ( 1 * 1 ) * 1 )} ] = broadcast( T14[ iS42{i3}, rS98{( ( i1 * i5 ) * i7 )} ] )
T16[ iS49{i3}, bS132{( ( 1 * 1 ) * 1 )} ]
   = T15[ iS46{i3}, bS128{( ( 1 * 1 ) * 1 )} ]
   / d22;
T21[ iS57{i3}, bS134{( ( 1 * 1 ) * 1 )} ]
   = T16[ iS49{i3}, bS132{( ( 1 * 1 ) * 1 )} ]
   + double(1.0000000000000001e-05);
T22[ iS61{i3}, bS110{( ( 1 * 1 ) * 1 )} ]
   = rsqrtf(T21[ iS57{i3}, bS134{( ( 1 * 1 ) * 1 )} ]);
T23[ iS66{i3}, iS108{( ( i1 * i5 ) * i7 )} ]
   = T12[ iS34{i3}, iS120{( ( i1 * i5 ) * i7 )} ]
   * T22[ iS61{i3}, bS110{( ( 1 * 1 ) * 1 )} ];
T24[ iS70{i9}, bS106{( ( 1 * 1 ) * 1 )} ] = broadcast( T1[ iS4{i9} ] )
T25[ iS74{i3}, iS112{( ( i1 * i5 ) * i7 )} ]
   = T23[ iS66{i3}, iS108{( ( i1 * i5 ) * i7 )} ]
   * T24[ iS70{i9}, bS106{( ( 1 * 1 ) * 1 )} ];
T26[ iS78{i11}, bS104{( ( 1 * 1 ) * 1 )} ] = broadcast( T2[ iS5{i11} ] )
T27[ iS82{i3}, iS102{( ( i1 * i5 ) * i7 )} ]
   = T25[ iS74{i3}, iS112{( ( i1 * i5 ) * i7 )} ]
   + T26[ iS78{i11}, bS104{( ( 1 * 1 ) * 1 )} ];
T29[ iS90{i3}, rS118{( ( 1 * 1 ) * 1 )} ] = reduction( T22[ iS61{i3}, bS110{( ( 1 * 1 ) * 1 )} ], op = add, initial value = double(0) )
d66 = d22 - 1;
T30[ rS41{i1}, iS42{i3}, rS43{i5}, rS44{i7} ]
   = T14[ iS42{i3}, rS98{( ( i1 * i5 ) * i7 )} ];
}
```

so
 ```
T30[ rS41{i1}, iS42{i3}, rS43{i5}, rS44{i7} ]
   = T14[ iS42{i3}, rS98{( ( i1 * i5 ) * i7 )} ];
```
is the cachefork, which was created because `T14` was an output of fusion, cf:
[https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp#L283-L290](https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp#L283-L290)

The problem is that we now have a `SET` op with both input and output having reduction axes because T14 is reduced output. This confuses root domain mapping.

**Comment by shmsong** [comment]
thoughts on potential fix:

The most correct way seems to be removing the reduce axes on `T30`, i.e. output of `SET`. `scheduleNormalization` as it is now probably wouldn't like it.

**Comment by naoyam** [comment]
`T30` does not only confuse root domain mapping. It is fundamentally invalid as it has reduction axes even though it is not an output tensor of reduction op.

@rdspring1 I have to admit I don't fully understand what `cache_fork` is meant to do, but shouldn't it use only non-reduction axes? More specifically, here:

https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/tensor_view.cpp#L842

Shouldn't it do `TensorDomain::noReductions(getRootDomain())` as in the other cache functions?

**Comment by rdspring1** [comment]
We save certain intermediate tensors for the backwards pass, which is re-read from global memory for subsequent uses.
`cache_fork` is used to separate the global write from the other uses.

We should remove the reduction axes from the cache.

However, this segmented fusion is incorrect because of a mistake in the user Fusion IR. It created an output that was not present in the original fusion. The entire fusion should be a single kernel.

**Comment by shmsong** [comment]
The FusionIR shown earlier was one of the segments. This was the pathological case we use for debugging the cache_fork issue and we knew this one didn't fuse into one kernel. 

```
Segmented_Fusion{ 
g{18}

g{8, 7}

g{19, 17, 16}

g{6, 5, 9, 27, 4, 3, 11, 12, 10, 0, 1, 2, 15, 22, 24, 26, 20, 14, 21, 13, 28, 23, 25}

e{ g{18}
 -> g{19, 17, 16}
(T19[ iS55{i15} ]) }

e{ g{8, 7}
 -> g{6, 5, 9, 27, 4, 3, 11, 12, 10, 0, 1, 2, 15, 22, 24, 26, 20, 14, 21, 13, 28, 23, 25}
(T10[ bS25{1}, bS26{1}, bS27{1}, iS28{i13} ]) }

e{ g{6, 5, 9, 27, 4, 3, 11, 12, 10, 0, 1, 2, 15, 22, 24, 26, 20, 14, 21, 13, 28, 23, 25}
 -> g{19, 17, 16}
(T14[ rS41{i1}, iS42{i3}, rS43{i5}, rS44{i7} ]) }

e{ g{6, 5, 9, 27, 4, 3, 11, 12, 10, 0, 1, 2, 15, 22, 24, 26, 20, 14, 21, 13, 28, 23, 25}
 -> g{19, 17, 16}
(d66) }

group details:

g{(pointwise)
T19[ iS55{i15} ]
   = T4[ iS7{i15} ]
   * double(0.90000000000000002);
}

g{(pointwise)
T10[ bS25{1}, bS26{1}, bS27{1}, iS28{i13} ] = broadcast( T9[ iS24{i13} ] )
 , T9[ iS24{i13} ]
   = T3[ iS6{i13} ]
   * double(0.90000000000000002);
}

g{(pointwise)
T20[ iS56{i15} ]
   = T19[ iS55{i15} ]
   + T18[ iS54{i3} ];
 , T18[ iS54{i3} ]
   = T17[ iS53{i3} ]
   * double(0.10000000000000001);
 , T17[ iS53{i3} ]
   = T14[ rS41{i1}, iS42{i3}, rS43{i5}, rS44{i7} ]
   / d66;
}

g{(normalization)
T8[ bS21{1}, iS20{i3}, bS22{1}, bS23{1} ]
   = T7[ bS17{1}, iS16{i3}, bS18{1}, bS19{1} ]
   * double(0.10000000000000001);
 , T7[ bS17{1}, iS16{i3}, bS18{1}, bS19{1} ]
   = T6[ bS12{1}, iS13{i3}, bS14{1}, bS15{1} ]
   / d22;
 , T11[ bS31{1}, iS30{i3}, bS32{1}, iS29{i13} ]
   = T10[ bS25{1}, bS26{1}, bS27{1}, iS28{i13} ]
   + T8[ bS21{1}, iS20{i3}, bS22{1}, bS23{1} ];
 , T28[ rS85{1}, iS86{i3}, rS87{1}, rS88{1} ] = reduction( T7[ bS17{1}, iS16{i3}, bS18{1}, bS19{1} ], op = add, initial value = double(0) )
 , T6[ bS12{1}, iS13{i3}, bS14{1}, bS15{1} ] = broadcast( T5[ rS8{i1}, iS9{i3}, rS10{i5}, rS11{i7} ] )
 , T5[ rS8{i1}, iS9{i3}, rS10{i5}, rS11{i7} ] = reduction( T0[ iS0{i1}, iS1{i3}, iS2{i5}, iS3{i7} ], op = add, initial value = double(0) )
 , T13[ iS37{i1}, iS38{i3}, iS39{i5}, iS40{i7} ]
   = T12[ iS33{i1}, iS34{i3}, iS35{i5}, iS36{i7} ]
   * T12[ iS33{i1}, iS34{i3}, iS35{i5}, iS36{i7} ];
 , T14[ rS41{i1}, iS42{i3}, rS43{i5}, rS44{i7} ] = reduction( T13[ iS37{i1}, iS38{i3}, iS39{i5}, iS40{i7} ], op = add, initial value = double(0) )
 , T12[ iS33{i1}, iS34{i3}, iS35{i5}, iS36{i7} ]
   = T0[ iS0{i1}, iS1{i3}, iS2{i5}, iS3{i7} ]
   - T7[ bS17{1}, iS16{i3}, bS18{1}, bS19{1} ];
 , d20 = double(1) * i1;
 , d21 = d20 * i5;
 , d22 = d21 * i7;
 , d66 = d22 - 1;
 , T23[ iS65{i1}, iS66{i3}, iS67{i5}, iS68{i7} ]
   = T12[ iS33{i1}, iS34{i3}, iS35{i5}, iS36{i7} ]
   * T22[ bS62{1}, iS61{i3}, bS63{1}, bS64{1} ];
 , T25[ iS73{i1}, iS74{i3}, iS75{i5}, iS76{i7} ]
   = T23[ iS65{i1}, iS66{i3}, iS67{i5}, iS68{i7} ]
   * T24[ bS69{1}, iS70{i9}, bS71{1}, bS72{1} ];
 , T27[ iS81{i1}, iS82{i3}, iS83{i5}, iS84{i7} ]
   = T25[ iS73{i1}, iS74{i3}, iS75{i5}, iS76{i7} ]
   + T26[ bS77{1}, iS78{i11}, bS79{1}, bS80{1} ];
 , T21[ bS58{1}, iS57{i3}, bS59{1}, bS60{1} ]
   = T16[ bS50{1}, iS49{i3}, bS51{1}, bS52{1} ]
   + double(1.0000000000000001e-05);
 , T16[ bS50{1}, iS49{i3}, bS51{1}, bS52{1} ]
   = T15[ bS45{1}, iS46{i3}, bS47{1}, bS48{1} ]
   / d22;
 , T22[ bS62{1}, iS61{i3}, bS63{1}, bS64{1} ]
   = rsqrtf(T21[ bS58{1}, iS57{i3}, bS59{1}, bS60{1} ]);
 , T15[ bS45{1}, iS46{i3}, bS47{1}, bS48{1} ] = broadcast( T14[ rS41{i1}, iS42{i3}, rS43{i5}, rS44{i7} ] )
 , T29[ rS89{1}, iS90{i3}, rS91{1}, rS92{1} ] = reduction( T22[ bS62{1}, iS61{i3}, bS63{1}, bS64{1} ], op = add, initial value = double(0) )
 , T24[ bS69{1}, iS70{i9}, bS71{1}, bS72{1} ] = broadcast( T1[ iS4{i9} ] )
 , T26[ bS77{1}, iS78{i11}, bS79{1}, bS80{1} ] = broadcast( T2[ iS5{i11} ] )
}

} //Segmented_Fusion
```

The segment at `T14`, aka. making it an output of one kernel and input of another, was because of these exprs that didn't have the same axis width:
```
g{(pointwise)
T20[ iS56{i15} ]
   = T19[ iS55{i15} ]
   + T18[ iS54{i3} ];
 , T18[ iS54{i3} ]
   = T17[ iS53{i3} ]
   * double(0.10000000000000001);
 , T17[ iS53{i3} ]
   = T14[ rS41{i1}, iS42{i3}, rS43{i5}, rS44{i7} ]
   / d66;
}
```

**Comment by rdspring1** [comment]
I was referring to the Fusion IR written by the user in the cpp test, which caused the fusion to be split incorrectly.

**Comment by naoyam** [comment]
PR to fix cache_fork: #714

**Comment by csarofeen** [comment]
Closed as fixed.
